From 3af3e6ec7008931bf88e18e6a828e6eda5e6f084 Mon Sep 17 00:00:00 2001
From: Alexandre Julliard <julliard@winehq.org>
Date: Mon, 16 Jun 2025 12:28:33 +0200
Subject: [PATCH 014/293] winegcc: Refuse to do non-PE builds on platforms that
 don't support it.

---
 tools/winegcc/winegcc.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/tools/winegcc/winegcc.c b/tools/winegcc/winegcc.c
index 402eb41e3f2..374f7542cf3 100644
--- a/tools/winegcc/winegcc.c
+++ b/tools/winegcc/winegcc.c
@@ -2068,6 +2068,10 @@ int main(int argc, char **argv)
 
     if (!file_align) file_align = section_align;
 
+    if (!is_pe && target.cpu != CPU_i386 && target.cpu != CPU_x86_64)
+        error( "Non-PE builds are not supported on this platform. You need to use something like '--target=%s-windows'.\n",
+               target.cpu == CPU_ARM ? "arm" : "aarch64" );
+
     if (!winebuild)
     {
         if (wine_objdir) winebuild = strmake( "%s/tools/winebuild/winebuild%s", wine_objdir, EXEEXT );
-- 
2.43.0

