From 46d873ea9ac17d41e27819ffcb432b43a8d7fc92 Mon Sep 17 00:00:00 2001
From: Esme Povirk <esme@codeweavers.com>
Date: Mon, 16 Jun 2025 20:29:04 +0000
Subject: [PATCH 033/293] gdiplus: Account for gdi32 clipping in
 GdipFillRegion.

---
 dlls/gdiplus/graphics.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dlls/gdiplus/graphics.c b/dlls/gdiplus/graphics.c
index 8fa0d689cbe..9bc5550d5a6 100644
--- a/dlls/gdiplus/graphics.c
+++ b/dlls/gdiplus/graphics.c
@@ -4833,6 +4833,11 @@ static GpStatus get_clipped_region_hrgn(GpGraphics* graphics, GpRegion* region,
         GdipDeleteRegion(device_region);
     }
 
+    if (status == Ok && graphics->gdi_clip)
+    {
+        CombineRgn(*hrgn, *hrgn, graphics->gdi_clip, RGN_AND);
+    }
+
     return status;
 }
 
-- 
2.43.0

