From b4df52fd617c5807969a26706e65d2e13d411cbe Mon Sep 17 00:00:00 2001
From: Dylan Donnell <dylan.donnell@student.griffith.ie>
Date: Sun, 15 Jun 2025 00:23:39 +0300
Subject: [PATCH 035/293] kernelbase: Allocate a new buffer for the module name
 in LoadLibraryExA.

---
 dlls/kernelbase/loader.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/dlls/kernelbase/loader.c b/dlls/kernelbase/loader.c
index 7afbe0460eb..6d651b0cc9f 100644
--- a/dlls/kernelbase/loader.c
+++ b/dlls/kernelbase/loader.c
@@ -533,9 +533,16 @@ HMODULE WINAPI DECLSPEC_HOTPATCH LoadLibraryW( LPCWSTR name )
 HMODULE WINAPI DECLSPEC_HOTPATCH LoadLibraryExA( LPCSTR name, HANDLE file, DWORD flags )
 {
     WCHAR *nameW;
+    HMODULE module;
+
+    /* A new allocation is necessary due to TP Shell Service
+     * calling LoadLibraryExA from an LdrLoadDll hook */
+    if (!(nameW = file_name_AtoW( name, TRUE ))) return 0;
 
-    if (!(nameW = file_name_AtoW( name, FALSE ))) return 0;
-    return LoadLibraryExW( nameW, file, flags );
+    module = LoadLibraryExW( nameW, file, flags );
+
+    HeapFree( GetProcessHeap(), 0, nameW );
+    return module;
 }
 
 
-- 
2.43.0

