From a26d1742930670a330c8327734b1e8eea1de1125 Mon Sep 17 00:00:00 2001
From: Nikolay Sivov <nsivov@codeweavers.com>
Date: Tue, 17 Jun 2025 19:28:15 +0200
Subject: [PATCH 058/293] bluetoothapis: Fix typo in a format string
 (Coverity).

Signed-off-by: Nikolay Sivov <nsivov@codeweavers.com>
---
 dlls/bluetoothapis/main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/bluetoothapis/main.c b/dlls/bluetoothapis/main.c
index a93b0ccf87e..acf763860ae 100644
--- a/dlls/bluetoothapis/main.c
+++ b/dlls/bluetoothapis/main.c
@@ -759,7 +759,7 @@ static void bluetooth_auth_wizard_ask_response( struct bluetooth_auth_wizard_lis
     else
     {
         resource_id = IDS_AUTH_WIZARD_UNNAMED_PAIR_REQUEST;
-        swprintf( name_or_addr, ARRAY_SIZE( name_or_addr ), L"%02X:02X:%02X:%02X:%02X:%02X", addr.rgBytes[0],
+        swprintf( name_or_addr, ARRAY_SIZE( name_or_addr ), L"%02X:%02X:%02X:%02X:%02X:%02X", addr.rgBytes[0],
                   addr.rgBytes[1], addr.rgBytes[2], addr.rgBytes[3], addr.rgBytes[4], addr.rgBytes[5] );
     }
     format_args[0] = (DWORD_PTR)name_or_addr;
-- 
2.43.0

