From b1b5315c1334bf96b99071a98beab530e06c58c9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Wed, 18 Jun 2025 10:23:54 +0200
Subject: [PATCH 062/293] include: Add APP_LOCAL_DEVICE_ID definition.

---
 include/windef.h | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/include/windef.h b/include/windef.h
index 12b3e5be57a..bb0722b57ba 100644
--- a/include/windef.h
+++ b/include/windef.h
@@ -158,6 +158,11 @@ typedef struct _RECTL
 
 typedef const RECTL *LPCRECTL;
 
+typedef struct APP_LOCAL_DEVICE_ID
+{
+    BYTE value[32];
+} APP_LOCAL_DEVICE_ID;
+
 /* DPI awareness */
 typedef enum DPI_AWARENESS
 {
-- 
2.43.0

