From daa42508b53e3495347f265711ffca2423ea525b Mon Sep 17 00:00:00 2001
From: Vibhav Pant <vibhavp@gmail.com>
Date: Wed, 18 Jun 2025 03:32:18 +0530
Subject: [PATCH 084/293] include: Add definitions for IGattDeviceService2.

---
 ...s.devices.bluetooth.genericattributeprofile.idl | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/include/windows.devices.bluetooth.genericattributeprofile.idl b/include/windows.devices.bluetooth.genericattributeprofile.idl
index 375fe39bea7..e71645486e8 100644
--- a/include/windows.devices.bluetooth.genericattributeprofile.idl
+++ b/include/windows.devices.bluetooth.genericattributeprofile.idl
@@ -35,6 +35,7 @@ namespace Windows.Devices.Bluetooth {
     typedef enum BluetoothCacheMode BluetoothCacheMode;
     typedef enum BluetoothError BluetoothError;
 
+    runtimeclass BluetoothLEDevice;
     runtimeclass BluetoothDeviceId;
 }
 
@@ -243,6 +244,19 @@ namespace Windows.Devices.Bluetooth.GenericAttributeProfile {
         [propget] HRESULT AttributeHandle([out, retval] UINT16 *value);
     }
 
+    [
+        contract(Windows.Foundation.UniversalApiContract, 1.0),
+        exclusiveto(Windows.Devices.Bluetooth.GenericAttributeProfile.GattDeviceService),
+        uuid(fc54520b-0b0d-4708-bae0-9ffd9489bc59)
+    ]
+    interface IGattDeviceService2 : IInspectable requires Windows.Devices.Bluetooth.GenericAttributeProfile.IGattDeviceService, Windows.Foundation.IClosable
+    {
+        [propget] HRESULT Device([out, retval] Windows.Devices.Bluetooth.BluetoothLEDevice **value);
+        [propget] HRESULT ParentServices([out, retval] Windows.Foundation.Collections.IVectorView<Windows.Devices.Bluetooth.GenericAttributeProfile.GattDeviceService *> **value);
+        HRESULT GetAllCharacteristics([out, retval] Windows.Foundation.Collections.IVectorView<Windows.Devices.Bluetooth.GenericAttributeProfile.GattCharacteristic *> **value);
+        HRESULT GetAllIncludedServices([out, retval] Windows.Foundation.Collections.IVectorView<Windows.Devices.Bluetooth.GenericAttributeProfile.GattDeviceService *> **value);
+    }
+
     [
         contract(Windows.Foundation.UniversalApiContract, 4.0),
         exclusiveto(Windows.Devices.Bluetooth.GenericAttributeProfile.GattDeviceService),
-- 
2.43.0

