From 31b9d0c0cc67b3a03340c33f0e9942bee94c431b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sat, 31 May 2025 12:43:36 +0200
Subject: [PATCH 113/293] winemac: Remove now unnecessary context sync.

---
 dlls/winemac.drv/opengl.c | 37 +------------------------------------
 1 file changed, 1 insertion(+), 36 deletions(-)

diff --git a/dlls/winemac.drv/opengl.c b/dlls/winemac.drv/opengl.c
index 11f90ff8ac1..17eacc5eab2 100644
--- a/dlls/winemac.drv/opengl.c
+++ b/dlls/winemac.drv/opengl.c
@@ -111,7 +111,6 @@ static const GLubyte *(*pglGetString)(GLenum name);
 static PFN_glGetIntegerv pglGetIntegerv;
 static void (*pglReadPixels)(GLint x, GLint y, GLsizei width, GLsizei height,
                              GLenum format, GLenum type, void *pixels);
-static void (*pglViewport)(GLint x, GLint y, GLsizei width, GLsizei height);
 
 
 struct color_mode {
@@ -1591,16 +1590,6 @@ static void make_context_current(struct macdrv_context *context, BOOL read)
 }
 
 
-/**********************************************************************
- *              sync_context
- */
-static void sync_context(struct macdrv_context *context)
-{
-    if (sync_context_rect(context))
-        make_context_current(context, FALSE);
-}
-
-
 /**********************************************************************
  *              set_swap_interval
  */
@@ -2093,7 +2082,6 @@ static BOOL macdrv_context_flush( void *private, HWND hwnd, HDC hdc, int interva
     struct macdrv_context *context = private;
 
     set_swap_interval(context, interval);
-    sync_context(context);
 
     if (skip_single_buffer_flushes)
     {
@@ -2163,23 +2151,6 @@ static void macdrv_glReadPixels(GLint x, GLint y, GLsizei width, GLsizei height,
 }
 
 
-/**********************************************************************
- *              macdrv_glViewport
- *
- * Hook into glViewport as an opportunity to update the OpenGL context
- * if necessary.  This is modeled after what Mesa GLX's Apple
- * implementation does.
- */
-static void macdrv_glViewport(GLint x, GLint y, GLsizei width, GLsizei height)
-{
-    struct macdrv_context *context = NtCurrentTeb()->glReserved2;
-
-    sync_context(context);
-    macdrv_update_opengl_context(context->context);
-    pglViewport(x, y, width, height);
-}
-
-
 /***********************************************************************
  *              macdrv_wglBindTexImageARB
  *
@@ -2836,7 +2807,6 @@ UINT macdrv_OpenGLInit(UINT version, const struct opengl_funcs *opengl_funcs, co
     LOAD_FUNCPTR(glGetIntegerv);
     LOAD_FUNCPTR(glGetString);
     LOAD_FUNCPTR(glReadPixels);
-    LOAD_FUNCPTR(glViewport);
     LOAD_FUNCPTR(glCopyColorTable);
 
     if (!init_gl_info())
@@ -2990,7 +2960,6 @@ static void *macdrv_get_proc_address(const char *name)
     if (!strcmp(name, "glCopyPixels")) return macdrv_glCopyPixels;
     if (!strcmp(name, "glGetString")) return macdrv_glGetString;
     if (!strcmp(name, "glReadPixels")) return macdrv_glReadPixels;
-    if (!strcmp(name, "glViewport")) return macdrv_glViewport;
 
     /* redirect some OpenGL extension functions */
     if (!strcmp(name, "glCopyColorTable")) return macdrv_glCopyColorTable;
@@ -3005,11 +2974,7 @@ static BOOL macdrv_swap_buffers(void *private, HWND hwnd, HDC hdc, int interval)
     TRACE("hdc %p context %p/%p/%p\n", hdc, context, (context ? context->context : NULL),
           (context ? context->cglcontext : NULL));
 
-    if (context)
-    {
-        set_swap_interval(context, interval);
-        sync_context(context);
-    }
+    if (context) set_swap_interval(context, interval);
 
     if (hwnd)
     {
-- 
2.43.0

