From 0657d263d2567aa0c4896039e315cbfe8b760637 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 17 Jun 2025 10:29:03 +0200
Subject: [PATCH 213/293] server: Remove unused atom pinned member.

---
 server/atom.c | 11 ++++-------
 1 file changed, 4 insertions(+), 7 deletions(-)

diff --git a/server/atom.c b/server/atom.c
index 03db48cfdd4..15c55d49813 100644
--- a/server/atom.c
+++ b/server/atom.c
@@ -51,7 +51,6 @@ struct atom_entry
     struct atom_entry *next;   /* hash table list */
     struct atom_entry *prev;   /* hash table list */
     int                count;  /* reference count */
-    short              pinned; /* whether the atom is pinned or not */
     atom_t             atom;   /* atom handle */
     unsigned short     hash;   /* string hash */
     unsigned short     len;    /* string len */
@@ -146,8 +145,8 @@ static void atom_table_dump( struct object *obj, int verbose )
     {
         struct atom_entry *entry = table->atoms[i];
         if (!entry) continue;
-        fprintf( stderr, "  %04x: ref=%d pinned=%c hash=%d \"",
-                 entry->atom, entry->count, entry->pinned ? 'Y' : 'N', entry->hash );
+        fprintf( stderr, "  %04x: ref=%d hash=%d \"",
+                 entry->atom, entry->count, entry->hash );
         dump_strW( entry->str, entry->len, stderr, "\"\"");
         fprintf( stderr, "\"\n" );
     }
@@ -208,7 +207,6 @@ static atom_t add_atom( struct atom_table *table, const struct unicode_str *str
             if ((entry->next = table->entries[hash])) entry->next->prev = entry;
             table->entries[hash] = entry;
             entry->count  = 1;
-            entry->pinned = 0;
             entry->hash   = hash;
             entry->len    = str->len;
             memcpy( entry->str, str->str, str->len );
@@ -228,8 +226,7 @@ static void delete_atom( struct atom_table *table, atom_t atom, int if_pinned )
 {
     struct atom_entry *entry = get_atom_entry( table, atom );
     if (!entry) return;
-    if (entry->pinned && !if_pinned) set_error( STATUS_WAS_LOCKED );
-    else if (!--entry->count)
+    if (!--entry->count)
     {
         if (entry->next) entry->next->prev = entry->prev;
         if (entry->prev) entry->prev->next = entry->next;
@@ -373,7 +370,7 @@ DECL_HANDLER(get_atom_information)
         {
             set_reply_data( entry->str, min( entry->len, get_reply_max_size() ));
             reply->count = entry->count;
-            reply->pinned = entry->pinned;
+            reply->pinned = 0;
             reply->total = entry->len;
         }
         else reply->count = -1;
-- 
2.43.0

