From ddc15b268365533cae7cc8108cb7114c00f41657 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Wed, 7 May 2025 11:59:53 +0200
Subject: [PATCH 244/293] server: Get rid of the console signaled flag.

---
 server/console.c | 14 ++------------
 1 file changed, 2 insertions(+), 12 deletions(-)

diff --git a/server/console.c b/server/console.c
index a8e14d70cc0..8f01311892f 100644
--- a/server/console.c
+++ b/server/console.c
@@ -54,7 +54,6 @@ struct console
 {
     struct object                obj;           /* object header */
     struct event_sync           *sync;          /* sync object for wait/signal */
-    int                          signaled;      /* is console signaled */
     struct thread               *renderer;      /* console renderer thread */
     struct screen_buffer        *active;        /* active screen buffer */
     struct console_server       *server;        /* console server object */
@@ -545,7 +544,6 @@ static struct object *create_console(void)
     if (!(console = alloc_object( &console_ops ))) return NULL;
     console->sync          = NULL;
     console->renderer      = NULL;
-    console->signaled      = 0;
     console->active        = NULL;
     console->server        = NULL;
     console->fd            = NULL;
@@ -1590,16 +1588,8 @@ DECL_HANDLER(get_next_console_request)
 
     if (!server->console->renderer) server->console->renderer = current;
 
-    if (!req->signal)
-    {
-        server->console->signaled = 0;
-        reset_sync( server->console->sync );
-    }
-    else if (!server->console->signaled)
-    {
-        server->console->signaled = 1;
-        signal_sync( server->console->sync );
-    }
+    if (!req->signal) reset_sync( server->console->sync );
+    else signal_sync( server->console->sync );
 
     if (req->read)
     {
-- 
2.43.0

