From 479f6b4ac064e306cfce687bc09aae6dfeea02e7 Mon Sep 17 00:00:00 2001
From: Charlotte Pabst <cpabst@codeweavers.com>
Date: Fri, 20 Jun 2025 22:38:54 +0200
Subject: [PATCH 274/293] mf: Don't forward thinning to clock.

---
 dlls/mf/session.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/dlls/mf/session.c b/dlls/mf/session.c
index 47760cc44a7..20401acc4b4 100644
--- a/dlls/mf/session.c
+++ b/dlls/mf/session.c
@@ -1578,7 +1578,7 @@ static void session_complete_rate_change(struct media_session *session)
     session->presentation.flags &= ~SESSION_FLAG_PENDING_RATE_CHANGE;
     session_set_presentation_clock(session);
 
-    hr = IMFRateControl_SetRate(session->clock_rate_control, session->presentation.thin,
+    hr = IMFRateControl_SetRate(session->clock_rate_control, FALSE,
             session->presentation.rate);
 
     param.vt = VT_R4;
@@ -4743,7 +4743,10 @@ static HRESULT WINAPI session_rate_control_GetRate(IMFRateControl *iface, BOOL *
 
     TRACE("%p, %p, %p.\n", iface, thin, rate);
 
-    return IMFRateControl_GetRate(session->clock_rate_control, thin, rate);
+    if (thin)
+        *thin = FALSE;
+
+    return IMFRateControl_GetRate(session->clock_rate_control, NULL, rate);
 }
 
 static const IMFRateControlVtbl session_rate_control_vtbl =
-- 
2.43.0

