From 56dc137257acf35d4055c1c313653baaf00592ea Mon Sep 17 00:00:00 2001
From: Hans Leidekker <hans@codeweavers.com>
Date: Mon, 23 Jun 2025 09:59:23 +0200
Subject: [PATCH 280/293] widl: Correct element type for interface signature.

---
 tools/widl/metadata.c | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/tools/widl/metadata.c b/tools/widl/metadata.c
index 9c8eee770ab..3f65888770a 100644
--- a/tools/widl/metadata.c
+++ b/tools/widl/metadata.c
@@ -1540,17 +1540,20 @@ static UINT make_type_sig( const type_t *type, BYTE *buf )
         len += make_type_sig( type_array_get_element_type(type), buf + len );
         break;
 
-    case TYPE_DELEGATE:
     case TYPE_INTERFACE:
+        buf[len++] = ELEMENT_TYPE_OBJECT;
+        break;
+
+    case TYPE_DELEGATE:
     case TYPE_RUNTIMECLASS:
-        buf[0] = ELEMENT_TYPE_CLASS;
-        len = encode_int( typedef_or_ref(TABLE_TYPEREF, type->md.ref), buf + 1 ) + 1;
+        buf[len++] = ELEMENT_TYPE_CLASS;
+        len += encode_int( typedef_or_ref(TABLE_TYPEREF, type->md.ref), buf + 1 );
         break;
 
     case TYPE_ENUM:
     case TYPE_STRUCT:
-        buf[0] = ELEMENT_TYPE_VALUETYPE;
-        len = encode_int( typedef_or_ref(TABLE_TYPEREF, type->md.ref), buf + 1 ) + 1;
+        buf[len++] = ELEMENT_TYPE_VALUETYPE;
+        len += encode_int( typedef_or_ref(TABLE_TYPEREF, type->md.ref), buf + 1 );
         break;
 
     case TYPE_BASIC:
-- 
2.43.0

